enhancements/dns/configurable-dns-pod-placement.md

```diff
@@ -0,0 +1,306 @@
+DaemonSet while keeping the blanket toleration on the DaemonSet that manages
+`/etc/hosts`.  Splitting the DaemonSet also makes it possible to enable use of a
+custom node selector on the CoreDNS DaemonSet.
+
+### Goals
+
+1. Separate CoreDNS and management of `/etc/hosts` into separate DaemonSets.
+2. Enable cluster administrators to control where the CoreDNS DaemonSet is scheduled.
+
+### Non-Goals
+
+1. Enable cluster administrators to control the placement of the DaemonSet that manages `/etc/hosts`.
+2. Enforce security policies.
+
+## Proposal
+
+This enhancement has two distinct parts.  First, the DNS operator, which manages
+the "dns-default" DaemonSet, is modified to manage an additional "node-resolver"
+DaemonSet, and the "dns-node-resolver" container, which manages `/etc/hosts`, is
+moved from the "dns-default" DaemonSet to a new "node-resolver" DaemonSet.  As
+part of this change, the toleration for all taints is removed from the
+"dns-default" DaemonSet.  From the cluster administrator's perspective, this
+DaemonSet split is an internal change.
+
+Second, a new API is provided to enable cluster administrators to specify the
+desired placement of the "dns-default" DaemonSet's Pods, which, due to the first
+change, only run CoreDNS and no longer must be scheduled to every node.  This
+new API is the user-facing part of this enhancement.
+
+The DNS operator API is extended by adding an optional `NodePlacement` field
+with type `DNSNodePlacement` to `DNSSpec`:
+
+```go
+// DNSSpec is the specification of the desired behavior of the DNS.
+type DNSSpec struct {
+    // ...
+	
+	// nodePlacement enables explicit control over the scheduling of DNS pods.
+	//
+	// If unset, defaults are used. See nodePlacement for more details.
+	//
+	// +optional
+	NodePlacement DNSNodePlacement `json:"nodePlacement,omitempty"`
+}
+```
+
+The `DNSNodePlacement` type has fields to specify a node selector and
+tolerations:
+
+```go
+// DNSNodePlacement describes node scheduling configuration for DNS pods.
+type DNSNodePlacement struct {
+	// nodeSelector is the node selector applied to DNS pods.
+	//
+	// If unset, the default is the following:
+	//
+	//   beta.kubernetes.io/os: linux
+	//
+	// If set, the specified selector is used and replaces the default.
+	//
+	// +optional
+	NodeSelector *metav1.LabelSelector `json:"nodeSelector,omitempty"`
+
+	// tolerations is a list of tolerations applied to DNS pods.
+	//
+	// The default is an empty list.
+	//
+	// See https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/
+	//
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+}
+```
+
+By default, DNS Pods run on untainted Linux nodes.  The `NodePlacement` field
+enables cluster administrators to specify alternative parameters.  For example,
+the following DNS specifies that DNS Pods should run only on "infra" nodes
+(i.e., nodes that have the "node-role.kubernetes.io/infra" label):
```

Comment:
Might be worth mentioning that this example assumes infra nodes have no infra node taints.

Reply:
Thanks!  I amended the example to include a toleration for any "node-role.kubernetes.io/infra" `NoSchedule` taint.  I don't know how prevalent the use of this taint is, but I gather that it is something that cluster administrators who configure infra nodes often use, and adding the toleration shouldn't cause any problems if the taint doesn't exist, so the example might as well include it.